peer/peer.go

```diff
@@ -0,0 +1,1942 @@
+	invMsgListeners         map[string]func(*Peer, *wire.MsgInv)
+	headersMsgListeners     map[string]func(*Peer, *wire.MsgHeaders)
+	getDataMsgListeners     map[string]func(*Peer, *wire.MsgGetData)
+	getBlocksMsgListeners   map[string]func(*Peer, *wire.MsgGetBlocks)
+	getHeadersMsgListeners  map[string]func(*Peer, *wire.MsgGetHeaders)
+	filterAddMsgListeners   map[string]func(*Peer, *wire.MsgFilterAdd)
+	filterClearMsgListeners map[string]func(*Peer, *wire.MsgFilterClear)
+	filterLoadMsgListeners  map[string]func(*Peer, *wire.MsgFilterLoad)
+	versionMsgListeners     map[string]func(*Peer, *wire.MsgVersion)
+}
+
+// String returns the peer's address and directionality as a human-readable
+// string.
+func (p *Peer) String() string {
+	return fmt.Sprintf("%s (%s)", p.addr, directionString(p.inbound))
+}
+
+// isKnownInventory returns whether or not the peer is known to have the passed
+// inventory.  It is safe for concurrent access.
+func (p *Peer) isKnownInventory(invVect *wire.InvVect) bool {
+	p.knownInvMutex.Lock()
+	defer p.knownInvMutex.Unlock()
+
+	if p.knownInventory.Exists(invVect) {
+		return true
+	}
+	return false
+}
+
+// UpdateLastBlockHeight updates the last known block for the peer. It is safe
+// for concurrent access.
+func (p *Peer) UpdateLastBlockHeight(newHeight int32) {
+	p.StatsMtx.Lock()
+	defer p.StatsMtx.Unlock()
+
+	log.Tracef("Updating last block height of peer %v from %v to %v",
+		p.addr, p.lastBlock, newHeight)
+	p.lastBlock = int32(newHeight)
+}
+
+// UpdateLastAnnouncedBlock updates meta-data about the last block sha this
+// peer is known to have announced. It is safe for concurrent access.
+func (p *Peer) UpdateLastAnnouncedBlock(blkSha *wire.ShaHash) {
+	p.StatsMtx.Lock()
+	defer p.StatsMtx.Unlock()
+
+	log.Tracef("Updating last blk for peer %v, %v", p.addr, blkSha)
+	p.lastAnnouncedBlock = blkSha
+}
+
+// AddKnownInventory adds the passed inventory to the cache of known inventory
+// for the peer.  It is safe for concurrent access.
+func (p *Peer) AddKnownInventory(invVect *wire.InvVect) {
+	p.knownInvMutex.Lock()
+	defer p.knownInvMutex.Unlock()
+
+	p.knownInventory.Add(invVect)
+}
+
+// Id returns the peer id.
+func (p *Peer) Id() int32 {
+	return p.id
+}
+
+// NA returns the peer network address.
+func (p *Peer) NA() *wire.NetAddress {
+	return p.na
+}
+
+// Addr returns the peer address.
+func (p *Peer) Addr() string {
+	return p.addr
+}
+
+// Filter returns the peer bloom filter.
+func (p *Peer) Filter() *bloom.Filter {
+	return p.filter
+}
+
+// Inbound returns whether the peer is inbound.
+func (p *Peer) Inbound() bool {
+	return p.inbound
+}
+
+// Persistent returns whether the peer is persistent.
+func (p *Peer) Persistent() bool {
+	return p.persistent
+}
+
+// RetryCount returns the number of times to retry peer connection.
+func (p *Peer) RetryCount() int64 {
+	return p.retryCount
+}
+
+// RequestQueue returns the peer request queue.
+func (p *Peer) RequestQueue() []*wire.InvVect {
```

Comment:
I believe I commented elsewhere too, but allowing direct access to the request queue like this and in `SetRequestQueue` smells like a layer violation that is ripe for issues.